### PR TITLE
Good grief reports and comment improvements

### PR DESCRIPTION
**Pull request by Slendy**
# Main features
This PR adds basic processing of good grief reports and displays them in a new 'reports' tab that is limited to admins only
Picture of website menu bar

The report viewer contains details about the report like the time it was submitted, who was in the player's lobby, the level that they were in, and the reason for the report. Reports can be deleted and upon deletion, all related hashes are also removed from the resource folder.
The viewer for the screenshot can highlight which person is which in the picture similar to how it works in normal photos as well as highlighting the region that the user selected
Picture of grief report menu

Picture of grief report region highlighting

Comments have been slightly revamped on the website side of things and now include a way for logged in users to post on levels and profiles and rate others comments
Picture of new comments

# Other bug fixes
* Fixed an [issue](https://wanderingdeveloper.medium.com/c-cautionary-tail-the-dangers-of-sha256-reuse-2b5bb9c6fde9) where reusing the SHA1 object would result in different hashes for the same file.
* Boo and meh hashes are properly serialized in the User object
* Previously if you accidentally typed a space at the beginning or end of the search bar it wouldn't get trimmed and you would get zero results. Now, all spaces are removed from user and photo searches because usernames can't contain spaces. In level searches only the leading and trailing whitespace is removed

**I would also recommend squashing all the commits I made whenever this gets merged.**

### CODE REVIEW

File: ProjectLighthouse/Controllers/Website/Admin/AdminReportController.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        foreach (string hash in hashes)
+        {
+            if (System.IO.File.Exists($"png{Path.DirectorySeparatorChar}{hash}"))
+            {
+                System.IO.File.Delete($"png{Path.DirectorySeparatorChar}{hash}");
+            }
+            if (System.IO.File.Exists($"r{Path.DirectorySeparatorChar}{hash}"))
+            {
+                System.IO.File.Delete($"r{Path.DirectorySeparatorChar}{hash}");
+            }
```
Comment (jvyden) [inline ProjectLighthouse/Controllers/Website/Admin/AdminReportController.cs:47]:
```suggestion
            if (System.IO.File.Exists(Path.Combine("png", $"{hash}.png")))
            {
                System.IO.File.Delete(Path.Combine("png", $"{hash}.png"));
            }
            if (System.IO.File.Exists(Path.Combine("r", hash)))
            {
                System.IO.File.Delete(Path.Combine("r", hash));
            }
```

File: ProjectLighthouse/Controllers/Website/Admin/AdminReportController.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        }
+        this.database.Reports.Remove(report);
+
+        await this.database.SaveChangesAsync();
+
+        return this.Redirect("~/reports/0");
```
Comment (jvyden) [inline ProjectLighthouse/Controllers/Website/Admin/AdminReportController.cs:53]: Administrator things should be under `/admin`.
```suggestion
        return this.Redirect("~/admin/reports/0");
```

File: ProjectLighthouse/Controllers/Website/SlotPageController.cs [hunk 3]
```diff
@@ -23,6 +23,32 @@ public SlotPageController(Database database)
...
 
+    [HttpGet("rateComment")]
+    public async Task<IActionResult> RateComment([FromRoute] int id, [FromQuery] int commentId, [FromQuery] int rating)
+    {
+        User? user = this.database.UserFromWebRequest(this.Request);
+        if (user == null) return this.Redirect("~/login");
+
+        await this.database.RateComment(user,
+            commentId,
+            rating);
```
Comment (jvyden) [inline ProjectLighthouse/Controllers/Website/SlotPageController.cs:34]: What happened here? Did our formatting rules screw this up or was this done manually?

File: ProjectLighthouse/Controllers/Website/SlotPageController.cs [hunk 4]
```diff
@@ -23,6 +23,32 @@ public SlotPageController(Database database)
...
+
+        return this.Redirect("~/slot/" + id + "#" + commentId);
```
Comment (jvyden) [inline ProjectLighthouse/Controllers/Website/SlotPageController.cs:36]:
```suggestion
        return this.Redirect($"~/slot/{id}#{commentId}");
```

File: ProjectLighthouse/Controllers/Website/UserPageController.cs [hunk 5]
```diff
@@ -31,6 +34,30 @@ public async Task<IActionResult> HeartUser([FromRoute] int id)
...
 
+    [HttpGet("rateComment")]
+    public async Task<IActionResult> RateComment([FromRoute] int id, [FromQuery] int? commentId, [FromQuery] int? rating)
+    {
+        User? user = this.database.UserFromWebRequest(this.Request);
+        if (user == null) return this.Redirect("~/login");
+
+        await this.database.RateComment(user, commentId.GetValueOrDefault(), rating.GetValueOrDefault());
+
+        return this.Redirect("~/user/" + id + "#" + commentId);
```
Comment (jvyden) [inline ProjectLighthouse/Controllers/Website/UserPageController.cs:45]:
```suggestion
        return this.Redirect($"~/user/{id}#{commentId}");
```

File: ProjectLighthouse/Helpers/HashHelper.cs [hunk 6]
```diff
@@ -71,7 +72,7 @@ public static async Task<string> ComputeDigest(string path, string authCookie, S
 
     public static string Sha1Hash(string str) => Sha1Hash(Encoding.UTF8.GetBytes(str));
 
-    public static string Sha1Hash(byte[] bytes) => BitConverter.ToString(sha1.ComputeHash(bytes)).Replace("-", "");
+    public static string Sha1Hash(byte[] bytes) => BitConverter.ToString(SHA1.Create().ComputeHash(bytes)).Replace("-","");
```
Comment (jvyden) [inline ProjectLighthouse/Helpers/HashHelper.cs:75]: Is there a problem with using `HashHelper.sha1` here?

File: ProjectLighthouse/Helpers/ImageHelper.cs [hunk 7]
```diff
@@ -18,7 +18,7 @@ public static bool LbpFileToPNG(byte[] data, string hash, LbpFileType type)
     {
         if (type != LbpFileType.Jpeg && type != LbpFileType.Png && type != LbpFileType.Texture) return false;
 
-        if (File.Exists($"png/{hash}.png")) return true;
+        if (File.Exists($"png{Path.DirectorySeparatorChar}{hash}.png")) return true;
```
Comment (jvyden) [inline ProjectLighthouse/Helpers/ImageHelper.cs:21]:
```suggestion
        if (File.Exists(Path.Combine("png", $"{hash}.png")) return true;
```

File: ProjectLighthouse/Pages/Layouts/BaseLayout.cshtml [hunk 8]
```diff
@@ -19,6 +19,7 @@
 
         @if (Model.User.IsAdmin)
         {
+            Model.NavigationItems.Add(new PageNavigationItem("Reports", "/reports/0", "exclamation circle"));
```
Comment (jvyden) [inline ProjectLighthouse/Pages/Layouts/BaseLayout.cshtml:22]: Hmm, I would have this somewhere in the admin panel.

Comment (jvyden) [inline ProjectLighthouse/Pages/Layouts/BaseLayout.cshtml:22]:
```suggestion
            Model.NavigationItems.Add(new PageNavigationItem("Reports", "/admin/reports/0", "exclamation circle"));
```

File: ProjectLighthouse/Pages/ReportsPage.cshtml [hunk 9]
```diff
@@ -0,0 +1,208 @@
+﻿@page "/reports/{pageNumber:int}"
```
Comment (jvyden) [inline ProjectLighthouse/Pages/ReportsPage.cshtml:1]:
```suggestion
﻿@page "/admin/reports/{pageNumber:int}"
```

File: ProjectLighthouse/Pages/ReportsPage.cshtml [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+</script>
+
+@if (Model.PageNumber != 0)
+{
+    <a href="/reports/@(Model.PageNumber - 1)@(Model.SearchValue.Length == 0 ? "" : "?name=" + Model.SearchValue)">Previous Page</a>
+}
+@(Model.PageNumber + 1) / @(Model.PageAmount)
+@if (Model.PageNumber < Model.PageAmount - 1)
+{
+    <a href="/reports/@(Model.PageNumber + 1)@(Model.SearchValue.Length == 0 ? "" : "?name=" + Model.SearchValue)">Next Page</a>
```
Comment (jvyden) [inline ProjectLighthouse/Pages/ReportsPage.cshtml:207]:
```suggestion
    <a href="/admin/reports/@(Model.PageNumber - 1)@(Model.SearchValue.Length == 0 ? "" : "?name=" + Model.SearchValue)">Previous Page</a>
}
@(Model.PageNumber + 1) / @(Model.PageAmount)
@if (Model.PageNumber < Model.PageAmount - 1)
{
    <a href="/admin/reports/@(Model.PageNumber + 1)@(Model.SearchValue.Length == 0 ? "" : "?name=" + Model.SearchValue)">Next Page</a>
```

File: ProjectLighthouse/Pages/ReportsPage.cshtml.cs [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+        this.SearchValue = name.Replace(" ", string.Empty);
+
+        this.ReportCount = await this.Database.Reports.Include(r => r.ReportingPlayer).CountAsync(r => r.ReportingPlayer.Username.Contains(this.SearchValue));
+
+        this.PageNumber = pageNumber;
+        this.PageAmount = Math.Max(1, (int) Math.Ceiling((double) this.ReportCount / ServerStatics.PageSize));
+
+        if (this.PageNumber < 0 || this.PageNumber >= this.PageAmount)
+            return this.Redirect($"/reports/{Math.Clamp(this.PageNumber, 0, this.PageAmount - 1)}");
```
Comment (jvyden) [inline ProjectLighthouse/Pages/ReportsPage.cshtml.cs:49]:
```suggestion
            return this.Redirect($"/admin/reports/{Math.Clamp(this.PageNumber, 0, this.PageAmount - 1)}");
```

File: ProjectLighthouse/Pages/SlotPage.cshtml.cs [hunk 12]
```diff
@@ -33,6 +36,15 @@ public async Task<IActionResult> OnGet([FromRoute] int id)
...
         this.Slot = slot;
 
+        if (this.User == null) return this.Page();
+        
+        foreach (Comment c in this.Comments)
+        {
+            Reaction? reaction = await this.Database.Reactions.FirstOrDefaultAsync(r =>
+                r.UserId == this.User.UserId && r.TargetId == c.CommentId);
+            if (reaction != null) c.YourThumb = reaction.Rating;
+        }
```
Comment (jvyden) [inline ProjectLighthouse/Pages/SlotPage.cshtml.cs:46]: I wouldn't bother doing this if comments are disabled.

File: ProjectLighthouse/Types/User.cs [hunk 13]
```diff
@@ -190,8 +190,8 @@ public string Serialize(GameVersion gameVersion = GameVersion.LittleBigPlanet1)
                       LbpSerializer.BlankElement("photos") +
                       LbpSerializer.StringElement("heartCount", this.Hearts) +
                       LbpSerializer.StringElement("yay2", this.YayHash) +
-                      LbpSerializer.StringElement("boo2", this.YayHash) +
-                      LbpSerializer.StringElement("meh2", this.YayHash);
+                      LbpSerializer.StringElement("boo2", this.BooHash) +
+                      LbpSerializer.StringElement("meh2", this.MehHash);
```
Comment (jvyden) [inline ProjectLighthouse/Types/User.cs:194]: God damnit.

**Review by jvyden — CHANGES_REQUESTED** [review]
Whew, this one's huge!

This is a pretty solid start, just a few nitpicks...

File: ProjectLighthouse/Helpers/HashHelper.cs (hunk 6 above)
Comment (jvyden) [reply ProjectLighthouse/Helpers/HashHelper.cs:75]: Might help if I actually read.

**Comment by jvyden** [comment]
I lied, lol. A bunch of nitpicks :trollface:

File: ProjectLighthouse/Controllers/Website/SlotPageController.cs (hunk 3 above)
Comment (Slendy) [reply ProjectLighthouse/Controllers/Website/SlotPageController.cs:34]: probably got formatted and I didn't notice it

File: ProjectLighthouse/Pages/Layouts/BaseLayout.cshtml (hunk 8 above)
Comment (Slendy) [reply ProjectLighthouse/Pages/Layouts/BaseLayout.cshtml]: Moved to the admin panel now, my original idea was that reports could be viewed by moderators so I put them in a more convenient place.

Comment (jvyden) [reply ProjectLighthouse/Pages/Layouts/BaseLayout.cshtml]: Ah, makes sense. Yeah, once a moderation role is implemented there should be a dedicated mod panel.

**Comment by Slendy** [comment]
thanks ci script very cool. Rerun the tests, they pass locally and on my forked repo.

**Comment by jvyden** [comment]
Yeah, they just do that cause of a race condition sometimes. I haven't looked into it.

**Review by jvyden — APPROVED** [review]

**Pull request merged**